plugins/scaleflux/sfx-nvme.c

```diff
@@ -0,0 +1,877 @@
+    SFX_LOG_QUAL                = 0xc4,
+    SFX_LOG_MISMATCHLBA         = 0xc5,
+    SFX_LOG_MEDIA               = 0xc6,
+    SFX_LOG_BBT                 = 0xc7,
+    SFX_LOG_IDENTIFY            = 0xcc,
+    SFX_FEAT_ATOMIC             = 0x01,
+};
+
+enum sfx_nvme_admin_opcode {
+    nvme_admin_query_cap_info   = 0xd3,
+    nvme_admin_change_cap       = 0xd4,
+    nvme_admin_sfx_set_features = 0xd5,
+    nvme_admin_sfx_get_features = 0xd6,
+};
+
+struct sfx_freespace_ctx
+{
+    __u64 free_space;
+    __u64 phy_cap;     /* physical capacity, in unit of sector */
+    __u64 phy_space;   /* physical space considering OP, in unit of sector */
+    __u64 user_space;  /* user required space, in unit of sector*/
+    __u64 hw_used;     /* hw space used in 4K */
+    __u64 app_written; /* app data written in 4K */
+};
+
+struct nvme_capacity_info {
+    __u64 lba_sec_sz;
+    __u64 phy_sec_sz;
+    __u64 used_space;
+    __u64 free_space;
+};
+#pragma pack(push,1)
+struct nvme_additional_smart_log_item {
+    uint8_t            key;
+    uint8_t            _kp[2];
+    uint8_t            norm;
+    uint8_t            _np;
+    union {
+        uint8_t        raw[6];
+        struct wear_level {
+            uint16_t    min;
+            uint16_t    max;
+            uint16_t    avg;
+        } wear_level ;
+        struct thermal_throttle {
+            uint8_t    pct;
+            uint32_t    count;
+        } thermal_throttle;
+    };
+    uint8_t            _rp;
+};
+#pragma pack(pop)
+
+struct nvme_additional_smart_log {
+    struct nvme_additional_smart_log_item    program_fail_cnt;
+    struct nvme_additional_smart_log_item    erase_fail_cnt;
+    struct nvme_additional_smart_log_item    wear_leveling_cnt;
+    struct nvme_additional_smart_log_item    e2e_err_cnt;
+    struct nvme_additional_smart_log_item    crc_err_cnt;
+    struct nvme_additional_smart_log_item    timed_workload_media_wear;
+    struct nvme_additional_smart_log_item    timed_workload_host_reads;
+    struct nvme_additional_smart_log_item    timed_workload_timer;
+    struct nvme_additional_smart_log_item    thermal_throttle_status;
+    struct nvme_additional_smart_log_item    retry_buffer_overflow_cnt;
+    struct nvme_additional_smart_log_item    pll_lock_loss_cnt;
+    struct nvme_additional_smart_log_item    nand_bytes_written;
+    struct nvme_additional_smart_log_item    host_bytes_written;
+    struct nvme_additional_smart_log_item    raid_recover_cnt; // errors which can be recovered by RAID
+    struct nvme_additional_smart_log_item    prog_timeout_cnt;
+    struct nvme_additional_smart_log_item    erase_timeout_cnt;
+    struct nvme_additional_smart_log_item    read_timeout_cnt;
+    struct nvme_additional_smart_log_item    read_ecc_cnt;//retry cnt
+};
+
+int nvme_change_cap(int fd, __u32 nsid, __u64 capacity)
+{
+    struct nvme_admin_cmd cmd = {
+    .opcode      = nvme_admin_change_cap,
+    .nsid	     = nsid,
+    };
+
+    cmd.cdw10 = (capacity & 0xffffffff);
```

Comment:
you can just set these in the struct declaration above instead of outside it.

Reply:
will change.